File: apps/web/src/lib/schemas.ts
/**
 * Zod schemas for everything crossing the network boundary.
 *
 * docs/06-frontend.md validates WebSocket frames with Zod at the boundary; the same
 * discipline is applied to REST responses here, so a server-side shape change surfaces as
 * a named parse error rather than as `undefined` three components deep.
 *
 * These mirror `apps/api/src/coach/services/models.py`.
 */

import { z } from 'zod';

export const taskStateSchema = z.enum([
  'draft',
  'not_started',
  'in_progress',
  'completed',
  'postponed',
  'postponed_until',
  'discarded',
]);
export type TaskState = z.infer<typeof taskStateSchema>;

export const researchStatusSchema = z.enum([
  'none',
  'pending',
  'in_progress',
  'done',
  'failed',
]);
export const originSchema = z.enum(['user', 'agent']);
export const projectStatusSchema = z.enum(['active', 'paused', 'archived']);
export const guidanceStyleSchema = z.enum(['socratic', 'direct', 'mixed']);
export type GuidanceStyle = z.infer<typeof guidanceStyleSchema>;

export const guidanceLevelSchema = z.enum(['mostly_guided', 'balanced', 'mostly_unguided']);
export type GuidanceLevel = z.infer<typeof guidanceLevelSchema>;

export const verbositySchema = z.enum(['terse', 'balanced', 'thorough']);
export type Verbosity = z.infer<typeof verbositySchema>;

export const researchDepthSchema = z.enum(['light', 'standard', 'deep']);
export type ResearchDepth = z.infer<typeof researchDepthSchema>;

// Moved above `taskItemSchema`, which now reuses it: `kind` was dropped when a report/
// proposed item was promoted into a checklist item, until this UI rework threaded it
// through end to end (`apps/api/src/coach/services/models.py`'s `TaskItem.kind`) so a
// checklist row can show the same kind chip its source item did.
export const reportItemKindSchema = z.enum([
  'article',
  'video',
  'exercise',
  'doc',
  'code_scaffold',
]);
export type ReportItemKind = z.infer<typeof reportItemKindSchema>;

export const taskItemSchema = z.object({
  itemId: z.string(),
  shortDescription: z.string(),
  details: z.string().default(''),
  //: `null` for a hand-added item, and for any item added before this field existed.
  kind: reportItemKindSchema.nullable().default(null),
  guided: z.boolean().default(false),
  completed: z.boolean().default(false),
  completedAt: z.string().nullable().default(null),
  minutes: z.number().int().nullable().default(null),
  url: z.string().nullable().default(null),
  sourceReportId: z.string().nullable().default(null),
});
export type TaskItem = z.infer<typeof taskItemSchema>;

export const rollupSchema = z.object({
  subtaskCount: z.number().int(),
  completedSubtasks: z.number().int(),
  totalEstimatedMinutes: z.number().int(),
});
export type Rollup = z.infer<typeof rollupSchema>;

export const taskSchema = z.object({
  id: z.string(),
  projectId: z.string(),
  ownerUid: z.string(),
  parentTaskId: z.string().nullable().default(null),
  title: z.string(),
  description: z.string().default(''),
  state: taskStateSchema,
  postponedUntil: z.string().nullable().default(null),
  estimatedMinutes: z.number().int().default(45),
  actualMinutes: z.number().int().nullable().default(null),
  order: z.string(),
  sessionId: z.string().nullable().default(null),
  needsResearch: z.boolean().default(true),
  researchStatus: z.enum(['none', 'pending', 'in_progress', 'done', 'failed']).default('none'),
  researchRequestedAt: z.string().nullable().default(null),
  latestReportId: z.string().nullable().default(null),
  studyPlanRunId: z.string().nullable().default(null),
  items: z.array(taskItemSchema).default([]),
  rollup: rollupSchema.nullable().default(null),
  origin: z.enum(['user', 'agent']).default('user'),
  createdAt: z.string().nullable().default(null),
  updatedAt: z.string().nullable().default(null),
  completedAt: z.string().nullable().default(null),
});
export type Task = z.infer<typeof taskSchema>;

export const taskWithSubtasksSchema = taskSchema.extend({
  subtasks: z.array(taskSchema).default([]),
});
export type TaskWithSubtasks = z.infer<typeof taskWithSubtasksSchema>;

export const projectPrefsSchema = z.object({
  defaultTaskMinutes: z.number().int().nullable().default(null),
  guidanceStyle: guidanceStyleSchema.nullable().default(null),
  guidanceLevel: guidanceLevelSchema.nullable().default(null),
  researchDepth: researchDepthSchema.nullable().default(null),
  allowVideos: z.boolean().nullable().default(null),
  confirmItemCompletion: z.boolean().nullable().default(null),
  preferredSources: z.array(z.string()).nullable().default(null),
  avoidSources: z.array(z.string()).nullable().default(null),
});
export type ProjectPrefs = z.infer<typeof projectPrefsSchema>;

export const projectCountsSchema = z.object({
  total: z.number().int(),
  completed: z.number().int(),
  openMinutes: z.number().int(),
});

export const roadmapBriefSchema = z.object({
  subject: z.string().default(''),
  specificTopics: z.array(z.string()).default([]),
  timeBudget: z.string().default(''),
  additionalNotes: z.string().default(''),
  attachments: z.array(z.string()).default([]),
  updatedAt: z.string().nullable().default(null),
});
export type RoadmapBrief = z.infer<typeof roadmapBriefSchema>;

export const projectSchema = z.object({
  id: z.string(),
  ownerUid: z.string(),
  title: z.string(),
  description: z.string().default(''),
  status: projectStatusSchema,
  prefs: projectPrefsSchema,
  nextUpTaskId: z.string().nullable().default(null),
  /**
   * The intake session `POST /api/projects` opens (docs/04-api-contract.md). A pointer on
   * the project so that the board can render the conversation without a round trip to
   * resolve it; `POST /api/projects/{id}/session` is the fallback for projects created
   * before it existed.
   */
  intakeSessionId: z.string().nullable().default(null),
  roadmapBrief: roadmapBriefSchema.nullable().default(null),
  counts: projectCountsSchema,
  lastAutonomousRunAt: z.string().nullable().default(null),
  createdAt: z.string().nullable().default(null),
  updatedAt: z.string().nullable().default(null),
});
export type Project = z.infer<typeof projectSchema>;

/** `POST /api/projects/{id}/troubleshooting/delete-all-tasks` — project settings'
 * troubleshooting action, not a board mutation. */
export const deleteAllTasksResponseSchema = z.object({
  deletedTasks: z.number().int(),
  resetPlans: z.number().int(),
});

export const effectivePrefsSchema = z.object({
  defaultTaskMinutes: z.number().int(),
  guidanceStyle: guidanceStyleSchema,
  guidanceLevel: guidanceLevelSchema.default('balanced'),
  verbosity: verbositySchema,
  timezone: z.string(),
  researchDepth: researchDepthSchema,
  allowVideos: z.boolean(),
  confirmItemCompletion: z.boolean(),
  preferredSources: z.array(z.string()),
  avoidSources: z.array(z.string()),
});
export type EffectivePrefs = z.infer<typeof effectivePrefsSchema>;

export const globalPrefsSchema = z.object({
  defaultTaskMinutes: z.number().int(),
  guidanceStyle: guidanceStyleSchema,
  verbosity: verbositySchema,
  timezone: z.string(),
  autonomousEnabled: z.boolean(),
  autonomousQuietHours: z.object({ start: z.string(), end: z.string() }),
});
export type GlobalPrefs = z.infer<typeof globalPrefsSchema>;

export const learnerProfileSchema = z.object({
  thinkingStyle: z.string().default(''),
  strengths: z.array(z.string()).default([]),
  gaps: z.array(z.string()).default([]),
  skills: z
    .array(
      z.object({
        name: z.string(),
        area: z.string().default('general'),
        level: z.string(),
        evidence: z.string().default(''),
      }),
    )
    .default([]),
  pacing: z.string().default(''),
  feedbackNotes: z.array(z.string()).default([]),
  updatedAt: z.string().nullable().default(null),
  updatedBy: z.enum(['agent', 'user']).default('user'),
  version: z.number().int().default(0),
});
export type LearnerProfile = z.infer<typeof learnerProfileSchema>;

export const learnerProfileResponseSchema = z.object({
  learnerProfile: learnerProfileSchema,
});

export const planLimitsSchema = z.object({
  autonomousRunsPerDay: z.number().int(),
  monthlyPoints: z.number().int(),
  fourHourPoints: z.number().int(),
});
export type PlanLimits = z.infer<typeof planLimitsSchema>;

/** One usage window (`GET /api/me`'s `usage.{monthly,fourHour}`), M8-quotas. */
export const usageWindowSchema = z.object({
  spent: z.number().int(),
  limit: z.number().int(),
  resetsAt: z.string(),
});
export type UsageWindow = z.infer<typeof usageWindowSchema>;

export const usageStatusSchema = z.object({
  monthly: usageWindowSchema,
  fourHour: usageWindowSchema,
});
export type UsageStatus = z.infer<typeof usageStatusSchema>;

export const meSchema = z.object({
  uid: z.string(),
  email: z.string().nullable(),
  displayName: z.string().nullable(),
  photoUrl: z.string().nullable(),
  globalPrefs: globalPrefsSchema,
  learnerProfile: learnerProfileSchema,
  plan: z.object({
    tier: z.string(),
    limits: planLimitsSchema,
  }),
  usage: usageStatusSchema,
});
export type Me = z.infer<typeof meSchema>;

export const couponClaimResponseSchema = z.object({
  plan: z.object({ tier: z.string(), limits: planLimitsSchema }),
});
export type CouponClaimResponse = z.infer<typeof couponClaimResponseSchema>;

export const projectListSchema = z.object({ projects: z.array(projectSchema) });
export const boardSchema = z.object({ tasks: z.array(taskWithSubtasksSchema) });
export const effectivePrefsResponseSchema = z.object({
  projectId: z.string(),
  effectivePrefs: effectivePrefsSchema,
});

export const projectDerivedSchema = z.object({
  id: z.string(),
  nextUpTaskId: z.string().nullable(),
  counts: z.record(z.string(), z.number()),
});

export const taskMutationSchema = z.object({
  task: taskSchema,
  parent: taskSchema.nullable().default(null),
  project: projectDerivedSchema.nullable().default(null),
});
export type TaskMutation = z.infer<typeof taskMutationSchema>;

// --- research reports ------------------------------------------------------------------

export const reportItemSchema = z.object({
  itemId: z.string(),
  kind: reportItemKindSchema,
  title: z.string(),
  url: z.string().nullable().default(null),
  minutes: z.number().int(),
  why: z.string().default(''),
  details: z.string().default(''),
  source: z.enum(['youtube', 'web', 'generated']).default('web'),
  meta: z.record(z.string(), z.string()).default({}),
  guided: z.boolean().nullable().default(null),
});
export type ReportItem = z.infer<typeof reportItemSchema>;

/**
 * `progress` holds feedback and nothing else. Per-item completion moved onto the task at
 * M4 (docs/02-data-model.md#task-items): a task has one checklist and may have several
 * reports, so completion on the report was answering the wrong question.
 */
export const researchReportSchema = z.object({
  id: z.string(),
  projectId: z.string(),
  ownerUid: z.string(),
  /** `null` since M8: a report from research about the project as a whole, kicked off
   * with no parent task — nothing is promoted into any task's checklist for one. */
  taskId: z.string().nullable().default(null),
  runId: z.string().nullable().default(null),
  sessionId: z.string().nullable().default(null),
  summary: z.string().default(''),
  required: z.array(reportItemSchema).default([]),
  optional: z.array(reportItemSchema).default([]),
  totalRequiredMinutes: z.number().int().default(0),
  budgetMinutes: z.number().int().default(45),
  citations: z.array(z.object({ uri: z.string(), title: z.string().default('') })).default([]),
  progress: z
    .object({ feedback: z.record(z.string(), z.enum(['up', 'down'])).default({}) })
    .default({ feedback: {} }),
  createdAt: z.string().nullable().default(null),
  updatedAt: z.string().nullable().default(null),
});
export type ResearchReport = z.infer<typeof researchReportSchema>;

export const reportResponseSchema = z.object({ report: researchReportSchema });

export const researchAcceptedSchema = z.object({
  runId: z.string(),
  turnId: z.string().nullable(),
  /** + M8: the run's own dedicated session — the research view watches this one, not
   * the session the request was made from. */
  sessionId: z.string(),
  mode: z.string(),
});

export const taskDetailSchema = z.object({
  task: taskWithSubtasksSchema,
  latestReport: researchReportSchema.nullable().default(null),
});
export type TaskDetail = z.infer<typeof taskDetailSchema>;

// --- autonomous runs -------------------------------------------------------------------
// docs/05-autonomous-runs.md#run-ledger. The banner reads `changes[]` and `undoneAt`; the
// run detail reads `steps[]`. Parsed loosely on purpose where the server owns the
// vocabulary — `status` and `step.id` are strings rather than enums here, because a run
// gaining a sixth step should not make an older tab fail to parse the ledger it is
// rendering. The UI switches on the values it knows and shows the rest as-is.

export const runStepSchema = z.object({
  id: z.string(),
  status: z.string(),
  startedAt: z.string().nullable().default(null),
  endedAt: z.string().nullable().default(null),
  output: z.record(z.string(), z.unknown()).nullable().default(null),
  error: z.string().nullable().default(null),
});
export type RunStep = z.infer<typeof runStepSchema>;

export const runChangeSchema = z.object({
  kind: z.string(),
  taskId: z.string(),
  previousOrder: z.string().nullable().default(null),
});
export type RunChange = z.infer<typeof runChangeSchema>;

export const autonomousRunSchema = z.object({
  id: z.string(),
  ownerUid: z.string(),
  projectId: z.string(),
  taskId: z.string().nullable().default(null),
  trigger: z.string(),
  mode: z.string(),
  status: z.string(),
  attempts: z.number().int().default(1),
  maxAttempts: z.number().int().default(3),
  steps: z.array(runStepSchema).default([]),
  turnId: z.string().nullable().default(null),
  /** + M8: the run's own dedicated research session, never the caller's. */
  sessionId: z.string().nullable().default(null),
  changes: z.array(runChangeSchema).default([]),
  undoneAt: z.string().nullable().default(null),
  createdAt: z.string().nullable().default(null),
  updatedAt: z.string().nullable().default(null),
  error: z.string().nullable().default(null),
});
export type AutonomousRun = z.infer<typeof autonomousRunSchema>;

export const runResponseSchema = z.object({ run: autonomousRunSchema });
export const runListSchema = z.object({ runs: z.array(autonomousRunSchema) });
export const runUndoSchema = z.object({
  run: autonomousRunSchema,
  taskIds: z.array(z.string()).default([]),
});

// --- study plans -------------------------------------------------------------------------
// docs/03-agent-design.md#the-taskless-case-task_proposer-and-plan_tailor-replace-reviewer_writer.
// A roadmap run's final document, read back from `GET /api/runs/{runId}/plan` —
// `get_run_report`'s sibling for `build_roadmap_workflow`. `proposedItemSchema` mirrors
// `reportItemSchema` minus `itemId`: a proposed item has none yet, it is only minted once
// `materialize_study_plan` promotes it onto a real task.

export const proposedItemSchema = z.object({
  kind: reportItemKindSchema,
  title: z.string(),
  url: z.string().nullable().default(null),
  minutes: z.number().int(),
  why: z.string().default(''),
  details: z.string().default(''),
  source: z.enum(['youtube', 'web', 'generated']).default('web'),
  guided: z.boolean().nullable().default(null),
});
export type ProposedItem = z.infer<typeof proposedItemSchema>;

export const proposedTaskSchema = z.object({
  slug: z.string(),
  title: z.string(),
  description: z.string().default(''),
  required: z.array(proposedItemSchema).default([]),
  optional: z.array(proposedItemSchema).default([]),
  prerequisiteTasks: z.array(z.string()).default([]),
});
export type ProposedTask = z.infer<typeof proposedTaskSchema>;

export const planDecisionSchema = z.enum(['include', 'additional', 'exclude', 'reject']);
export type PlanDecision = z.infer<typeof planDecisionSchema>;

export const planTaskEntrySchema = z.object({
  taskSlug: z.string(),
  after: z.string().nullable().default(null),
  prerequisiteTasks: z.array(z.string()).default([]),
  relevance: z.number().int().min(0).max(4).default(0),
  decision: planDecisionSchema,
  why: z.string().default(''),
});
export type PlanTaskEntry = z.infer<typeof planTaskEntrySchema>;

export const studyPlanSchema = z.object({
  id: z.string(),
  projectId: z.string(),
  ownerUid: z.string(),
  runId: z.string().nullable().default(null),
  sessionId: z.string().nullable().default(null),
  title: z.string().default(''),
  shortDescription: z.string().default(''),
  longDescription: z.string().default(''),
  memo: z.string().default(''),
  proposedTasks: z.array(proposedTaskSchema).default([]),
  plan: z.array(planTaskEntrySchema).default([]),
  materializedAt: z.string().nullable().default(null),
  materializedTaskIds: z.array(z.string()).default([]),
  createdAt: z.string().nullable().default(null),
  updatedAt: z.string().nullable().default(null),
});
export type StudyPlan = z.infer<typeof studyPlanSchema>;

export const studyPlanResponseSchema = z.object({ plan: studyPlanSchema });

// --- sessions & turns ------------------------------------------------------------------

export const sessionSummarySchema = z.object({
  id: z.string(),
  projectId: z.string().nullable().default(null),
  taskId: z.string().nullable().default(null),
});
export type SessionSummary = z.infer<typeof sessionSummarySchema>;

export const sessionResponseSchema = z.object({ session: sessionSummarySchema });

/**
 * One transcript row. `event` is the serialized ADK `Event` verbatim — the API returns it
 * unprojected on purpose (docs/02-data-model.md), so the parts this UI actually renders
 * are picked out by `transcript.ts` rather than pinned by a schema that would have to
 * track a pinned dependency's model.
 */
export const sessionEventSchema = z.object({
  seq: z.number().int(),
  eventId: z.string(),
  event: z.record(z.string(), z.unknown()),
});
export type SessionEvent = z.infer<typeof sessionEventSchema>;

export const sessionEventsSchema = z.object({
  events: z.array(sessionEventSchema),
  nextAfterSeq: z.number().int(),
  hasMore: z.boolean(),
});

export const turnStatusSchema = z.enum(['running', 'complete', 'failed', 'cancelled']);
export type TurnStatus = z.infer<typeof turnStatusSchema>;

export const turnAcceptedSchema = z.object({
  turnId: z.string(),
  sessionId: z.string(),
  status: turnStatusSchema,
  startSeq: z.number().int().default(0),
});

export const turnStatusResponseSchema = z.object({
  turnId: z.string(),
  status: turnStatusSchema,
  lastSeq: z.number().int(),
});

export const wsTicketSchema = z.object({ ticket: z.string(), expiresAt: z.string() });

export const uploadCreatedSchema = z.object({ uploadId: z.string(), signedUrl: z.string() });
export const uploadFinalizedSchema = z.object({ uploadId: z.string(), mimeType: z.string() });

/** RFC 9457 problem+json, which is how every API error arrives. */
export const problemSchema = z.object({
  type: z.string().default('about:blank'),
  title: z.string().default('Error'),
  status: z.number().int(),
  detail: z.string().default(''),
  instance: z.string().optional(),
  /**
   * Present on the `409` from `POST /api/sessions/{sid}/research`: the run already holding
   * the project's agent lease. What makes that response actionable rather than a dead end
   * (docs/04-api-contract.md#post-apisessionssidresearch).
   */
  runId: z.string().optional(),
  /**
   * Present on the `429` from a quota-exceeded turn (M8-quotas): which window
   * (`"monthly" | "4-hour"`) and when it resets, in addition to `detail`'s
   * human-readable sentence carrying the same information
   * (docs/04-api-contract.md#usage-quotas-implemented-m8-quotas).
   */
  window: z.string().optional(),
  resetAt: z.string().optional(),
});
export type Problem = z.infer<typeof problemSchema>;

File: apps/api/src/coach/api/schemas.py
"""Request and response bodies.

Wire shapes from docs/04-api-contract.md. Responses reuse the domain models directly —
they already serialize to camelCase — so there is no second definition of a task to keep
in step. Requests get their own models because they are patches and partials, where the
difference between "absent" and "null" matters.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from coach.services.models import (
    AutonomousRun,
    EffectivePrefs,
    GlobalPrefs,
    GuidanceLevel,
    GuidanceStyle,
    ItemFeedback,
    LearnerProfile,
    Minutes,
    Plan,
    Project,
    ProjectStatus,
    ResearchDepth,
    ResearchReport,
    SessionSummary,
    Task,
    TaskState,
    TaskWithSubtasks,
    TurnStatus,
    UsageStatus,
    Verbosity,
)


class RequestModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, extra="forbid")


class ResponseModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


# --- identity & preferences ----------------------------------------------------------


class MeResponse(ResponseModel):
    uid: str
    email: str | None
    display_name: str | None
    photo_url: str | None
    global_prefs: GlobalPrefs
    learner_profile: LearnerProfile
    plan: Plan
    #: docs/02-data-model.md#usage-quotas-m8-quotas. Spend, limit, and reset time for both
    #: windows, regardless of whether either is exhausted.
    usage: UsageStatus


class MeIdentityPatch(RequestModel):
    """`PATCH /api/me`. Only `displayName` today — the one identity field a learner may
    override; `email` and `photoUrl` stay Identity Platform's to set."""

    display_name: str = Field(min_length=1, max_length=100)


class CouponClaimRequest(RequestModel):
    code: str = Field(min_length=1, max_length=64)


class CouponClaimResponse(ResponseModel):
    plan: Plan


class QuietHoursPatch(RequestModel):
    start: str
    end: str


class GlobalPrefsPatch(RequestModel):
    """`PATCH /api/me/prefs`. Every field optional; absent means "leave alone"."""

    default_task_minutes: Minutes | None = None
    guidance_style: GuidanceStyle | None = None
    verbosity: Verbosity | None = None
    timezone: str | None = None
    autonomous_enabled: bool | None = None
    autonomous_quiet_hours: QuietHoursPatch | None = None


class TechnologyBeliefPatch(RequestModel):
    name: str
    level: str
    evidence: str = ""


class LearnerProfilePatch(RequestModel):
    thinking_style: str | None = Field(default=None, max_length=500)
    strengths: list[str] | None = None
    gaps: list[str] | None = None
    technologies: list[TechnologyBeliefPatch] | None = None
    pacing: str | None = None
    feedback_notes: list[str] | None = None


# --- projects ------------------------------------------------------------------------


class ProjectCreate(RequestModel):
    title: str = Field(min_length=1, max_length=200)
    goal: str = ""


class ProjectPrefsPatch(RequestModel):
    default_task_minutes: Minutes | None = None
    guidance_style: GuidanceStyle | None = None
    guidance_level: GuidanceLevel | None = None
    research_depth: ResearchDepth | None = None
    allow_videos: bool | None = None
    confirm_item_completion: bool | None = None
    preferred_sources: list[str] | None = None
    avoid_sources: list[str] | None = None


class ProjectPatch(RequestModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    goal: str | None = None
    status: ProjectStatus | None = None
    prefs: ProjectPrefsPatch | None = None


class ProjectListResponse(ResponseModel):
    projects: list[Project]


class BoardResponse(ResponseModel):
    """`GET /api/projects/{id}/tasks` — parents with nested subtasks."""

    tasks: list[TaskWithSubtasks]


# --- tasks ---------------------------------------------------------------------------


class TaskCreate(RequestModel):
    title: str = Field(min_length=1, max_length=300)
    description: str = ""
    estimated_minutes: Minutes | None = None
    parent_task_id: str | None = None
    after_task_id: str | None = None
    needs_research: bool = True


class TaskPatch(RequestModel):
    title: str | None = Field(default=None, min_length=1, max_length=300)
    description: str | None = None
    estimated_minutes: Minutes | None = None
    needs_research: bool | None = None


class TaskStateChange(RequestModel):
    state: TaskState
    postponed_until: datetime | None = None


class TaskReorder(RequestModel):
    """Exactly one of the two must be given; the service enforces that."""

    after_task_id: str | None = None
    before_task_id: str | None = None


# `SubtaskDraft`/`TaskSplit` and `POST /api/tasks/{id}/split` were removed after M4. A
# subtask is now created by `POST /api/projects/{id}/tasks` with a `parentTaskId`, one at a
# time — see the note in `services/tasks.py` for why the all-at-once shape had to go.


class TaskItemDraft(RequestModel):
    """One checklist item, as a client or a tool supplies it.

    No `itemId`: it is assigned server-side, like a report item's
    (docs/02-data-model.md#task-items). No `completed` either — an item is added as
    outstanding work, and ticking it is a `PATCH` on the item it became.
    """

    short_description: str = Field(min_length=1, max_length=300)
    details: str = ""
    guided: bool = False
    minutes: Minutes | None = None
    url: str | None = None


class TaskItemsAdd(RequestModel):
    items: list[TaskItemDraft] = Field(min_length=1, max_length=30)


class TaskItemPatch(RequestModel):
    completed: bool | None = None
    short_description: str | None = Field(default=None, min_length=1, max_length=300)
    details: str | None = None
    guided: bool | None = None


class TaskItemReorder(RequestModel):
    """Exactly one of the two must be given; the service enforces that."""

    after_item_id: str | None = None
    before_item_id: str | None = None


class ProjectDerived(ResponseModel):
    """The project fields a task mutation can move."""

    id: str
    next_up_task_id: str | None
    counts: dict[str, int]


class TaskMutationResponse(ResponseModel):
    """What every task mutation returns.

    docs/04-api-contract.md: "Reorder and state changes return the full updated task
    (plus the affected parent) so the client can reconcile optimistically without a
    refetch." `project` is carried alongside for the same reason — `counts` and
    `nextUpTaskId` move on the same write, and without them the board's "Next up" pin
    would need a second round-trip to settle.
    """

    task: Task
    parent: Task | None = None
    project: ProjectDerived | None = None


class TaskDetailResponse(ResponseModel):
    task: TaskWithSubtasks
    #: docs/04-api-contract.md: `GET /api/tasks/{id}` returns "Task + `items[]` + subtasks
    #: + `latestReport`". The checklist is on the task; the report is the material behind
    #: it, and the workspace renders `optional[]` and the citations from here.
    latest_report: ResearchReport | None = None


class RunResponse(ResponseModel):
    """`GET /api/runs/{runId}` — the whole ledger row.

    Whole rather than projected: the client renders `steps[]` as progress, keys the
    "Updated by your coach" banner off `changes[]` and `undoneAt`, and reads `turnId` to
    attach to the stream. Picking a subset here would mean a second endpoint the first
    time the UI wanted one more field.
    """

    run: AutonomousRun


class RunListResponse(ResponseModel):
    runs: list[AutonomousRun]


class RunUndoResponse(ResponseModel):
    """`POST /api/runs/{runId}/undo`.

    `taskIds` is what the client invalidates. It is returned rather than inferred from
    `run.changes` because undo tolerates a task that has already gone, so what it *did*
    touch is a strictly smaller list than what the run changed.
    """

    run: AutonomousRun
    task_ids: list[str]


class ReportListResponse(ResponseModel):
    """`GET /api/tasks/{id}/reports` — newest first.

    A list because reports accumulate rather than replacing each other
    (docs/10-risks.md Q4); the UI renders the newest and collapses the rest.
    """

    reports: list[ResearchReport]


class ReportItemFeedback(RequestModel):
    """`PATCH /api/reports/{reportId}/items/{itemId}`.

    Feedback and nothing else. This body used to carry `completed` as well; item completion
    moved onto the task at M4 (docs/04-api-contract.md#tasks), and the field is *absent*
    rather than ignored so a client still sending one gets a 422 instead of a silent
    success. `taskId` is required because a report is addressed without its project and
    ownership is checked through the task.
    """

    task_id: str
    feedback: ItemFeedback | None = None


class ReportResponse(ResponseModel):
    report: ResearchReport


class ResearchRequest(RequestModel):
    reason: str = Field(default="", max_length=2000)
    budget_minutes_override: Minutes | None = None
    #: Re-run even when the task already has materials.
    force: bool = False
    #: + M8: files the learner attached to the question, forwarded into the research
    #: session's opening turn exactly as `TurnRequest.attachments` are — an upload is
    #: addressable by `uploadId` from any session, so this needs no new plumbing beyond
    #: accepting the same shape here.
    attachments: list[TurnAttachment] = Field(default_factory=list)


class ResearchResponse(ResponseModel):
    """The 202 from `POST /api/sessions/{sid}/research`.

    `turnId` is what the client subscribes to. `sessionId` (+ M8) is the run's own fresh
    session — never the `sid` the request was made against — and is what the research view
    reads to render that run's transcript (docs/04-api-contract.md#post-apisessionssidresearch).
    """

    run_id: str
    turn_id: str | None
    session_id: str
    mode: str


class EffectivePrefsResponse(ResponseModel):
    project_id: str
    effective_prefs: EffectivePrefs


# --- sessions & turns ------------------------------------------------------------------


class SessionResponse(ResponseModel):
    session: SessionSummary


class SessionEventView(ResponseModel):
    """One transcript row.

    `event` is the serialized ADK `Event` verbatim, not a projection of it. The transcript
    is ADK's data (docs/02-data-model.md nests the whole event under `event_data`), and
    re-shaping it here would mean a second definition of a conversation turn that has to
    be kept in step with a pinned dependency's model.
    """

    seq: int
    event_id: str
    event: dict[str, Any]


class SessionEventsResponse(ResponseModel):
    events: list[SessionEventView]
    next_after_seq: int
    has_more: bool


class TurnAttachment(RequestModel):
    upload_id: str
    mime_type: str


class TurnConfirmation(RequestModel):
    """The learner's answer to a tool that asked before acting.

    `discard_task` is gated by ADK's `require_confirmation` (docs/03-agent-design.md), so
    the turn that proposes it ends with an `adk_request_confirmation` function call and
    resumes only when a matching function *response* arrives. This is that response,
    shaped as fields rather than as a raw ADK part: the client should not be building ADK
    payloads, and the call id is the only thing it has to carry back.

    `payload` carries a *structured* answer, which is what makes `ask_learner` possible:
    ADK's `ToolConfirmation` has a free-form `payload` beside `confirmed`, so a tool can
    ask a question rather than only for approval and get the selection back through the
    same handshake. It is `None` for the yes/no gates, whose whole answer is `confirmed`.

    Untyped here on purpose. The shape belongs to the tool that asked — `ask_learner`
    validates the selection against the options it offered, which is where the check has
    to be anyway, since the payload has been through the client.
    """

    function_call_id: str
    confirmed: bool
    payload: dict[str, Any] | None = None


class TurnRequest(RequestModel):
    """`POST /api/sessions/{sid}/turns`.

    `idempotencyKey` is in the contract's example body; it is accepted here *and* as the
    `Idempotency-Key` header, which is what `IdempotencyMiddleware` actually reads. The
    body field is kept so a client following the contract's example is not silently
    unprotected.
    """

    text: str = ""
    attachments: list[TurnAttachment] = Field(default_factory=list)
    confirmation: TurnConfirmation | None = None
    idempotency_key: str | None = None


class TurnAcceptedResponse(ResponseModel):
    turn_id: str
    session_id: str
    status: TurnStatus
    start_seq: int = 0


class TurnStatusResponse(ResponseModel):
    turn_id: str
    status: TurnStatus
    last_seq: int


class WsTicketResponse(ResponseModel):
    ticket: str
    expires_at: datetime


# --- uploads ---------------------------------------------------------------------------


class UploadCreate(RequestModel):
    filename: str = Field(min_length=1, max_length=255)
    mime_type: str
    size_bytes: int = Field(gt=0)


class UploadCreated(ResponseModel):
    upload_id: str
    signed_url: str


class UploadFinalized(ResponseModel):
    upload_id: str
    mime_type: str
